lib/ansible/plugins/loader.py

```diff
@@ -269,16 +285,19 @@ def add_directory(self, directory, with_subdir=False):
         ''' Adds an additional directory to the search path '''
 
         directory = os.path.realpath(directory)
+        orig_size = len(self._extra_dirs)
 
         if directory is not None:
             if with_subdir:
                 directory = os.path.join(directory, self.subdir)
             if directory not in self._extra_dirs:
                 # append the directory and invalidate the path cache
                 self._extra_dirs.append(directory)
-                self._paths = None
                 display.debug('Added %s to loader search path' % (directory))
 
+        if orig_size != len(self._extra_dirs):
+            self._clear_caches()
```

Comment:
Why not just do this unconditionally after the `append()` call above?

Reply:
was thinking of accepting multiple dirs .. but that is probably premature